File: python/day15.py
import threading
import time

def worker(id):
    print(f"worker {id} starting....")
    time.sleep(1)
    print(f"Worker {id} done.")

# Main execution
# Lets store thread objects in list
threads = []
new_workers = 3

for i in range(1, new_workers + 1):
    t = threading.Thread(target=worker, args=(i,))
    t.start()
    threads.append(t)

print("Main is waiting for threads to finish....")

# Loop over list of threads and join() each other.
# This blocks the main thread, waiting for 't' to finish
# before moving to the next one.
for t in threads:
    t.join()

print("All threads finish. Main is done.")

File: go-basic/day15.go
// In Day 14 we saw Go's famous feature Concurrency using Goroutine.
// Incredibly lightweight thread managed by go runtime.
// previously we use time.sleep() to wait for goroutines.(bad way)

// Today we will use sync.WaitGroup() counter to wait for goroutine to finish.

package main

import (
	"fmt"
	"sync"
	"time"
)

// Create a worker function which accepts a pointer to the sync.WaitGroup()
// so it ca call Done()
func worker(id int, wg *sync.WaitGroup) {
	// We will use special keyword 'defer' which says run this line of code at very end of this function.
	// Right before it returns.
	// This guarantees Done() is called even if the function had an error.
	defer wg.Done()
	fmt.Printf("Worker %d starting......\n", id)

	// Simulate some work
	time.Sleep(1 * time.Second)

	fmt.Printf("Worker %d done. \n", id)
}

func main() {
	// Create a new WaitGroup
	var wg sync.WaitGroup

	newWorkers := 3

	// tell WaitGroup that we are waiting for 3 goroutines
	wg.Add(newWorkers)

	// lets launch 3 goroutines
	for i := 1; i <= newWorkers; i++ {
		go worker(i, &wg)
	}
	fmt.Println("Main is waiting for workers to finish....")
	wg.Wait()
	fmt.Println("All workers finished. Main is exiting.")
}
